wasm/jit/jit_amd64.go

```diff
@@ -4594,66 +4569,695 @@ func (c *amd64Compiler) allocateRegister(t generalPurposeRegisterType) (reg int1
+		readCompiledFunctionAddress.To.Type = obj.TYPE_REG
+		readCompiledFunctionAddress.To.Reg = tmpRegister
+		readCompiledFunctionAddress.From.Type = obj.TYPE_MEM
+		readCompiledFunctionAddress.From.Reg = callFrameStackTopAddressRegister
+		// "rc.caller" is BELOW the top address. See the above example for detail.
+		readCompiledFunctionAddress.From.Offset = -(callFrameDataSize - callFrameCompiledFunctionOffset)
+		c.addInstruction(readCompiledFunctionAddress)
+
+		// At this point, tmpRegister holds the caller's *compiledFunction,
+		// so we are ready to read the address of the caller's module instance
+		// stored at compiledFunction.ModuleInstanceAddress.
+		readModuleInstanceAddress := c.newProg()
+		readModuleInstanceAddress.As = x86.AMOVQ
+		readModuleInstanceAddress.To.Type = obj.TYPE_REG
+		readModuleInstanceAddress.To.Reg = moduleInstanceAddressRegister
+		readModuleInstanceAddress.From.Type = obj.TYPE_MEM
+		readModuleInstanceAddress.From.Reg = tmpRegister
+		readModuleInstanceAddress.From.Offset = compiledFunctionModuleInstanceAddressOffset
+		c.addInstruction(readModuleInstanceAddress)
+
+		putModuleInstanceAddressToEngine := c.newProg()
+		putModuleInstanceAddressToEngine.As = x86.AMOVQ
+		putModuleInstanceAddressToEngine.To.Type = obj.TYPE_MEM
+		putModuleInstanceAddressToEngine.To.Reg = reservedRegisterForEngine
+		putModuleInstanceAddressToEngine.To.Offset = engineModuleContextModuleInstanceAddressOffset
+		putModuleInstanceAddressToEngine.From.Type = obj.TYPE_REG
+		putModuleInstanceAddressToEngine.From.Reg = moduleInstanceAddressRegister
+		c.addInstruction(putModuleInstanceAddressToEngine)
+	}
+
+	// Complete 3).
+	{
+		readReturnAddress := c.newProg()
+		readReturnAddress.As = x86.AMOVQ
+		readReturnAddress.To.Type = obj.TYPE_REG
+		readReturnAddress.To.Reg = tmpRegister
+		readReturnAddress.From.Type = obj.TYPE_MEM
+		readReturnAddress.From.Reg = callFrameStackTopAddressRegister
+		// "ra.caller" is BELOW the top address. See the above example for detail.
+		readReturnAddress.From.Offset = -(callFrameDataSize - callFrameReturnAddressOffset)
+		c.addInstruction(readReturnAddress)
+
+		jmpToReturnAddress := c.newProg()
+		jmpToReturnAddress.As = obj.AJMP
+		jmpToReturnAddress.To.Type = obj.TYPE_REG
+		jmpToReturnAddress.To.Reg = tmpRegister
+		c.addInstruction(jmpToReturnAddress)
+	}
+
+	// They were temporarily used, so we mark them unused.
+	c.locationStack.markRegisterUnused(regs...)
+	return nil
+}
 
-	// Release all the registers as our calling convention requires the callee-save.
+// callGoFunction adds instructions to call a Go function whose address equals the addr parameter.
+// jitStatus is set before making call, and it should be either jitCallStatusCodeCallBuiltInFunction or
+// jitCallStatusCodeCallHostFunction.
+func (c *amd64Compiler) callGoFunction(jitStatus jitCallStatusCode, addr wasm.FunctionAddress) error {
+	// Set the functionAddress to the engine.exitContext.functionCallAddress.
```

Comment:
double nested usually doesn't resolve

```suggestion
	// Set the functionAddress to the engine.exitContext functionCallAddress.
```